### PR TITLE
#293: Auto-update of configured status

### PR DESCRIPTION
**Pull request by katiefenn**
Fixes: #293 
- Added polled update of configured status to dashboard

### CODE REVIEW

File: src/common/actions/snaps.js [hunk 1]
```diff
@@ -12,14 +12,16 @@ export const REMOVE_SNAP = 'REMOVE_SNAP';
 export const REMOVE_SNAP_SUCCESS = 'REMOVE_SNAP_SUCCESS';
 export const REMOVE_SNAP_ERROR = 'REMOVE_SNAP_ERROR';
 
-export function fetchUserSnaps(owner) {
+export function fetchUserSnaps(owner, showFetching = true) {
```
Comment (katiefenn) [inline src/common/actions/snaps.js]: New param stops fetching status (i.e. the spinner) from being shown while fetching new snap data. This is useful when polling the service for updates.

File: src/common/components/repositories-home/index.js [hunk 2]
```diff
@@ -33,12 +36,22 @@ class RepositoriesHome extends Component {
...
     const owner = this.props.user.login;
 
     if (authenticated) {
-      this.props.dispatch(fetchUserSnaps(owner));
+      dispatch(fetchUserSnaps(owner));
+      if (!interval) {
+        interval = setInterval(() => {
+          dispatch(fetchUserSnaps(owner, false));
+        }, SNAP_POLL_PERIOD);
+      }
```
Comment (katiefenn) [inline src/common/components/repositories-home/index.js]: Poll `fetchUserSnaps` with a fifteen second interval, only if an existing timer has not been set up.

File: src/common/components/repositories-home/index.js [hunk 3]
```diff
@@ -33,12 +36,22 @@ class RepositoriesHome extends Component {
...
     }
+  }
 
+  componentWillUnmount() {
+    clearInterval(interval);
+    interval = null;
```
Comment (katiefenn) [inline src/common/components/repositories-home/index.js]: Clear timer, preventing snap data being fetched when user navigates away from the page.

File: src/common/actions/snaps.js (hunk 1 above)
Comment (earnubs) [inline src/common/actions/snaps.js]: How do you check if you're fetching?

Comment (bartaz) [inline src/common/actions/snaps.js]: Not sure if I agree with such approach. This makes the state in store little bit inconsistent, because if we call it with `showFetching = false` data is fetching (we have request in progress), but the store doesn't reflect this change.

If we don't want to show spinner during polling (and I agree with that) we should make UI make this decision based on state in the store - not make the store pretend there is no request in progress.

In build pages (where we poll for builds statuses) it's done as `showLoading = isFetching && !success` - but this assumes that action that inits fetching doesn't clear the success status in store. This is not the case in `snaps` reducer as far as I can see. And I guess it's debatable what should be our convention in such case (should we clear success when we set isFetching)?

Other option would be do do `showLoading = isFetching && !snaps`. So we will only show loading if we don't have snaps data already.

So, to summarize my point here: I think the state should always have isFetching set when we wait for response (for example so we can detect it and not fetch again until completed), and it's UI component that should decide about showing loading based on what is in the store. I don't think we should create 'silent' action that fetch requests, but hide some info from the store.

Comment (earnubs) [inline src/common/actions/snaps.js]: Another option is that the UI always shows the isFetching state but in a less obtrusive way, for example if the spinner is behind the content (only visible when we have no content). Cheating, I know.

File: src/common/components/repositories-home/index.js [hunk 4]
```diff
@@ -89,10 +89,11 @@ class RepositoriesHome extends Component {
...
   render() {
     const { snaps } = this.props;
+    const hasSnaps = (snaps.snaps && Object.keys(snaps.snaps).length > 0);
     // show spinner until we know if user has any enabled repos
-    return (snaps.success || snaps.error)
-        ? this.renderRepositoriesList()
-        : this.renderSpinner();
+    return (snaps.isFetching && !hasSnaps)
+      ? this.renderSpinner()
+      : this.renderRepositoriesList();
```
Comment (katiefenn) [inline src/common/components/repositories-home/index.js]: I've changed the criteria for showing the repositories list to avoid explicitly overriding the "isFetching" status.

`snaps.success` is not a reliable criteria, because it changes whenever we poll the service for updated snap information. Instead, I've used the `isFetching` status coupled with the existence of snaps.

File: test/unit/src/common/components/repositories-home/t_repositories-home.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import expect from 'expect';
+import { shallow } from 'enzyme';
+import proxyquire from 'proxyquire';
+import sinon from 'sinon';
+
+const RepositoriesHome = proxyquire(
+  '../../../../../../src/common/components/repositories-home',
+  {
+    '../../actions/snaps': {
+      fetchUserSnaps: sinon.stub().returns({ type: 'FETCH_SNAP_REPOSITORIES' })
```
Comment (bartaz) [inline test/unit/src/common/components/repositories-home/t_repositories-home.js]: I guess that's where `mapDispatchToProps` would help as suggested by @earnubs.

We would be able to pass mocked dispatch calls to raw component via props without ugly proxyquire calls.

File: src/common/components/repositories-home/index.js [hunk 6]
```diff
@@ -109,3 +123,4 @@ function mapStateToProps(state) {
 }
 
 export default connect(mapStateToProps)(withRouter(RepositoriesHome));
+export const raw = RepositoriesHome;
```
Comment (bartaz) [inline src/common/components/repositories-home/index.js]: I know we didn't settle down on any convention for this, but is basically the same as exporting `RepositoriesHome` directly - maybe just under different name, like `RepositoriesHomeView`.

But `RepositoriesHomeRaw` works for me as well. Simply - I would be for exporting it under full uppercase class name. It doesn't matter much in this case with proxyquire, but it does with import.

File: src/common/components/repositories-list/index.js [hunk 7]
```diff
@@ -100,15 +99,11 @@ class RepositoriesList extends Component {
...
 
   render() {
-    const isLoading = this.props.snaps.isFetching;
+    const { snaps } = this.props.snaps;
+    const hasSnaps = (snaps && Object.keys(snaps).length > 0);
 
     return (
       <div className={styles.repositoriesList}>
-        { isLoading &&
-          <div className={ spinnerStyles }>
```
Comment (bartaz) [inline src/common/components/repositories-list/index.js]: I guess this throws on linter, as `spinnerStyles` are still imported but not used.

File: test/unit/src/common/components/repositories-home/t_repositories-home.js (hunk 5 above)
Comment (katiefenn) [inline test/unit/src/common/components/repositories-home/t_repositories-home.js]: Yep, that would be good.

File: src/common/components/repositories-list/index.js (hunk 7 above)
Comment (katiefenn) [inline src/common/components/repositories-list/index.js]: This should now be resolved.

File: src/common/components/repositories-home/index.js (hunk 4 above)
Comment (earnubs) [inline src/common/components/repositories-home/index.js]: Why not an explicit poll action, with associated state?

Comment (katiefenn) [inline src/common/components/repositories-home/index.js]: @earnubs: It was something I considered, but as @bartaz suggested we already have state in the store that can tell us when we're polling and when we're fetching for the initial status. In fact, we could probably write a selector to do this for us.
